shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -40,10 +68,75 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]]) extends ZioTransEx
    * @param req
    *   the request (isomorphic to the Unit)
    */
-  def openConnection(req: ConnectRequest): ZIO[ZEnv, Status, Connection] = for {
+  def openConnection(req: ConnectRequest): Result[Connection] = for {
     id <- ZIO.effectTotal(UUID.randomUUID())
     _ <- addConnection(Conn(id))
   } yield Connection(id.toString())
 
-  private def addConnection(c: Conn): UIO[Unit] = connections.update(_ + (c.id -> c))
+  /**
+   * Parses a spec into a model
+   *
+   * This method handles the [LoadModelRequest] RPC defined by `transExplorer.proto`
+   *
+   * @param req
+   *   the request to load a model, including the root module spec and any auxiliary modules
+   */
+  def loadModel(req: LoadModelRequest): Result[LoadModelResponse] = for {
+    parseResult <- parseSpec(req.spec, req.aux)
+    result <- parseResult match {
+      case Right(module) =>
+        for {
+          _ <- updateConnection(req.conn)(_.setModel(module))
+          json = jsonOfModule(module)
+        } yield LoadModelResponse.Result.Spec(json)
+      case Left(err) =>
+        ZIO.succeed(LoadModelResponse.Result.Err(err.getMessage()))
+    }
+  } yield LoadModelResponse(result)
+
+  private def parseSpec(spec: String, aux: Seq[String]): Result[Either[Throwable, TlaModule]] =
+    // Obtain permit on the semaphore protecting access to the parser, ensuring the parser is not
+    // run by more than one thread at a time.
+    parserSemaphore.withPermit(ZIO.effectTotal(try {
+      val parser = Executor(new ParserModule)
+      parser.passOptions.set("parser.source", SourceOption.StringSource(spec, aux))
+      parser.run().left.map(code => ParsingFailed(s"Parsing failed with error code: ${code}"))
+    } catch {
+      case e: Throwable => Left(ParsingFailed(s"Parsing failed with exception: ${e.getMessage()}"))
+    }))
+
+  private def jsonOfModule(module: TlaModule): String = {
+    new TlaToUJson(None).makeRoot(Seq(module)).toString
+  }
+
+  private def addConnection(c: Conn): Result[Unit] = connections.update(_ + (c.id -> c))
+
+  private def getConnection(reqConn: RequestConn): Result[Conn] = {
+    for {
+      uuid <-
+        try {
+          reqConn match {
+            case Some(conn) => ZIO.succeed(UUID.fromString(conn.id))
+            case None       => throw new IllegalArgumentException()
+          }
+        } catch {
+          case _: IllegalArgumentException =>
+            // TODO log for invalid conn ID: https://github.com/informalsystems/apalache/issues/1849
+            ZIO.fail(Status.INVALID_ARGUMENT)
```

Comment:
Can we immediately return a `ZIO.fail` here?

Reply:
I meant in the `None` branch above, instead of throwing and catching `IllegalArgumentException`.

Reply:
Ah. Yeah we can do, I just figured this way had less repetition and gave us just two exit points, rather than three. iiuc, you think just returning the same fail on line 120 would read better?

Reply:
For me yes, but resolve it as you see fit :)

Reply:
Changed this.